File: Leetcode Challenge/February/cpp/Permutation in String.cpp
class Solution {
public:
    
    bool checkInclusion(string s1, string s2) {
        int m[26] = {0}; //character array
        for(char c : s1) m[c - 'a']++;
        int j = 0, i = 0, total_chars = s1.size();
        while(j < s2.size()){
            if(m[s2.at(j++) - 'a']-- > 0)
                total_chars--;
            if(total_chars == 0) return true;
            //shifting of window
            if(j - i == s1.size() && m[s2.at(i++) - 'a']++ >= 0)
                total_chars++;
        }
        return false;
    }
};

File: Leetcode Challenge/February/java/Permutation in String.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int[] map = new int[26];
        for(char c : s1.toCharArray()) map[c - 'a']++;
        int j = 0, i = 0;
        int count_chars = s1.length();
        while(j < s2.length()){
            if(map[s2.charAt(j++) - 'a']-- > 0)
                count_chars--;
            if(count_chars == 0) return true;
            if(j - i == s1.length() && map[s2.charAt(i++) - 'a']++ >= 0)
                count_chars++;
        }
        return false;
    }
}
